### ISSUE TITLE
Azure AD With Federated Workload Identity for Grafana Login

### ISSUE THREAD

**Issue by lynkz-matt-psaltis** [grafana/grafana#78249]
**Why is this needed**:
Currently a number of the LGTM components support authentication through Azure Workload Identity however it seems the main Grafana login process still requires a ClientId & Secret.

Adding Azure Workload Identity would help to prevent credential leak and secret expiry / rotation.

**What would you like to be added**:

Please consider adding support for this in the main grafana oauth component https://github.com/grafana/grafana/blob/main/pkg/services/authn/clients/oauth.go

Azure Workload Identity is already supported for data sources so hopefully its something that can be more easily applied to the login flow as well.

**Who is this feature for?**
Grafana dashboard users who utilise Azure AD for authentication

**Comment by taleena** [comment]
Additional information from a duplicate issue:

We already support both [managed identity](https://grafana.com/docs/grafana/latest/datasources/azure-monitor/#configure-managed-identity) and [workload identity](https://grafana.com/docs/grafana/latest/datasources/azure-monitor/#configure-workload-identity) credentials in the Azure Monitor data source.

This request applies to step 8 from this [setup guide](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/azuread/#create-the-microsoft-entra-id-application).

**Comment by prasadrajesh** [comment]
Hi @taleena , I think existing support is for only allowing integration Grafana with Azure Monitor data source through "Azure Workload Identity". But I think the ask was Grafana SSO login with Azure AD through "Azure Workload Identity" instead "Azure App Registration client and secret".

**Comment by tarunfeb27** [comment]
Hi, I am trying to configure Grafana AzureAD integration using the [setup guide](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/azuread/#create-the-microsoft-entra-id-application). mentioned by @taleena. I chose Federated credentials instead of Client Secerts but it fails to exchange token. Anyone was able to make it work ? 
```logger=authn.service t=2025-02-26T14:50:31.397076941Z level=error msg="Failed to authenticate request" client=auth.client.azuread error="[auth.oauth.token.exchange] failed to exchange code to token: oauth2: \"invalid_client\" \"AADSTS7000218: The request body must contain the following parameter: 'client_assertion' or 'client_secret'. ```

**Comment by mehighlow** [comment]
@tarunfeb27 , nope. Same issue. I created federated credentials, but according to the [configuration-options](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/azuread/#configuration-options) the `client_id` is still required. However, the key point of this open issue is that federated credentials replace the need for a password (client secret), making the setup more secure.

**Comment by mehighlow** [comment]
@tarunfeb27 , upon ~further investigation~ reading [documentation](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/azuread/#create-the-microsoft-entra-id-application) thoroughly, acording to section 'Federated credentials':
 "You must have set client_authentication under [auth.azuread] to managed_identity in the Grafana server configuration for this to work." 

In case your are soursing configuration from env vars, `GF_AUTH_AZUREAD_CLIENT_AUTHENTICATION=managed_identity`. Also, I removed `GF_AUTH_AZUREAD_CLIENT_SECRET`

**Comment by tarunfeb27** [comment]
@mehighlow  Thanks ! Yeah, I am at exactly same outcome. 
Looks like Grafana is not able to point to the manage identity/workload identity used in the OAuth config to exchange tokens.
Though the this [documentation](https://grafana.com/docs/grafana/v11.5/datasources/azure-monitor/#configure-workload-identity) relates to datasources, I tried to configure : ( _you might have tried that as well_ ) 
```[azure]
workload_identity_enabled = true
workload_identity_tenant_id = IDENTITY_TENANT_ID
workload_identity_client_id = IDENTITY_CLIENT_ID
workload_identity_token_file = TOKEN_FILE_PATH
```

But no luck yet.

**Comment by mehighlow** [comment]
@taleena , @kalleep , @tonypowa , could you please check this issue?

**Comment by taleena** [comment]
Hi all, you are right, I have been working more closely with the data sources side of things so my answer was with that context, whoops!

This issue has the right visibility from the team who would involved in this work. It looks like it's in "ToDo", so perhaps @kalleep or another can take it from here, and I'm happy to help where needed.

**Comment by mikemadeja** [comment]
The documentation seems to claim it exists. Terraform doesn't have the parameter that the documentation claims to have. 

https://github.com/grafana/terraform-provider-grafana/issues/2102

Based on the documentation, it seems to exist but doesn't work? I followed it and I get.

```
logger=authn.service t=2025-03-31T14:44:22.156968352Z level=error msg="Failed to authenticate request" client=auth.client.azuread error="[auth.oauth.token.exchange] failed to exchange code to token: error getting managed identity token: ManagedIdentityCredential
authentication failed. the requested identity isn't assigned to this resource\nGET http://169.254.169.254/metadata/identity/oauth2/token\n--------------------------------------------------------------------------------\nRESPONSE 400: 400 Bad
Request\n--------------------------------------------------------------------------------\n{\n  \"error\": \"invalid_request\",\n  \"error_description\": \"Identity not found\"\n}\n--------------------------------------------------------------------------------\nTo
troubleshoot, visit https://aka.ms/azsdk/go/identity/troubleshoot#managed-id"
```
I believe I am getting this when I log in. Here's a screenshot.

![Image](https://github.com/user-attachments/assets/349af1fe-f699-4a5c-9d32-13bcc3df6f4c)

**Comment by kheiakiyama** [comment]
Hello I updated to latest version, then it works!

I changed following parameters 

```
grafana.ini:
  auth.azuread:
    client_authentication: managed_identity
    managed_identity_client_id: "{GUID from AKS.}" # my case properties.identityProfile.kubeletidentity.clientId
    federated_credential_audience: "api://AzureADTokenExchange"
```

### Does work
Grafana version:11.6.0
Helm chart version: 8.11.0
K8s version: 1.30.7

### Does not work
Grafana version:11.3.1
Helm chart version: 8.6.3
K8s version: 1.30.7

**Comment by mehighlow** [comment]
@kheiakiyama ,

- Are you sharing the kubelet's managed identity (must be user-assigned, as only user-assigned managed identities can be added as federated credentials in Entra ID applications) identity with Grafana?
- Did you create Federated credentials (your managed identity and app registration)?
- What did you set as a client_id in your setup?


I tried `11.6.0`, but still observing the same issue I shared [earlier](https://github.com/grafana/grafana/issues/78249#issuecomment-2689526763)

**Comment by srakows** [comment]
I am experiencing the same issue. 

I think the confusion above is vmss managed identity (https://learn.microsoft.com/en-us/azure/aks/use-managed-identity), which uses the identity assigned to your entire nodepool(s), vs workload identity (https://learn.microsoft.com/en-us/azure/aks/workload-identity-overview), which assigns an identity to a specific pod. @kheiakiyama mentions `kubeletidentity` in his comment, which is the former. 

Further, the Grafana docs are wrong. This section (https://grafana.com/docs/grafana/v11.6/datasources/azure-monitor/#configure-workload-identity) explicitly refers to workload identity, but in the error messages above, we see `GET http://169.254.169.254/metadata/identity/oauth2/token`. This means Grafana is trying to contact the instance metadata service (https://learn.microsoft.com/en-us/azure/virtual-machines/instance-metadata-service), which will NOT work for a workload identity -- this service only applies to vmss-assigned managed identities (that is, identities assigned to the VM itself, not the workload). In a workload identity scenario, tokens are mounted in the filesystem (rather than being retrievable from a well-known uri), with the path (and tenant, and client id) indicated by env vars. See https://learn.microsoft.com/en-us/azure/aks/workload-identity-overview?tabs=go#how-it-works

**Comment by mehighlow** [comment]
@srakows , 

It doesn’t matter, in the context of this issue, which managed identity is used. One may choose (though this is far from the best option, if not the worst) to use a managed identity assigned to every VM in the VMSS (known as the kubelet’s identity) s long as it is a [user assigned identity](https://learn.microsoft.com/en-us/azure/aks/use-managed-identity#use-a-pre-created-kubelet-managed-identity) because [the documentation](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/azuread/#create-the-microsoft-entra-id-application) states:

<img width="764" alt="Image" src="https://github.com/user-attachments/assets/37260167-6a4e-457e-bda7-8be505791869" />

Also, this issue is not about Azure Monitor data source, but [Grafana OAuth authentication through Azure Entra ID(Azure AD](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/azuread/#create-the-microsoft-entra-id-application)).

Azure Monitor data source using managed identity works just fine.

**Comment by srakows** [comment]
You're right @mehighlow , I was looking at the wrong documentation. The documentation for Entra ID authentication doesn't actually mention workload identities. 

Everything works fine for us if we use the vmss managed identity, but we can't use that in production for obvious reasons. Specifically, using the auto-created user assigned identity in AKS (should be named `CLUSTER_NAME-agentpool`). 

To answer your earlier questions:

* this is using the vmss managed identity
* yes, you create the federated credential in the app registration, following https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation-config-app-trust-managed-identity?tabs=microsoft-entra-admin-center
* `client_id` is the app registration's client id

Specifically, these are the lines I added to the mounted grafana.ini:

```
[auth.azuread]
client_id = APP_REGISTRATION_CLIENT_ID
managed_identity_client_id = MSI_CLIENT_ID
client_authentication = managed_identity
federated_credential_audience = "api://AzureADTokenExchange"
```

This is on Grafana `11.5.2`

**Comment by kheiakiyama** [comment]
@mehighlow Sorry for your confusion.
@srakows Yes, your comment is correct. I was using the vmss managed identity.
For convenience, this was a different topic than Workload Identity. Sorry.
